packages/flutter/lib/src/rendering/paragraph.dart

```diff
@@ -95,10 +143,44 @@ class RenderParagraph extends RenderBox {
     span?.recognizer?.addPointer(event);
   }
 
+  bool _hasVisualOverflow = false;
+  TextPainter _overflowPainter;
+  ui.Shader _overflowShader;
+
   @override
   void performLayout() {
     _layoutText(constraints);
     size = constraints.constrain(_textPainter.size);
+
+    final bool didOverflowWidth = size.width < _textPainter.width;
+    _hasVisualOverflow = didOverflowWidth || size.height < _textPainter.height;
```

Comment:
No, `textPainter.height` just reflects the line boxes.  I'll add a TODO about this issue.